### ISSUE TITLE
move Sage conversion into Sage

### ISSUE THREAD

**Issue by rwst** [sympy/sympy#13430]
After the Sage ticket
https://trac.sagemath.org/ticket/24006

 - the `_sage_` member functions
 - the file `sympy/external/tests/test_sage.py`
 - Travis tests and scripts with `TEST_SAGE`

can be removed.

Motivation:
The real problems of #13395 raise the question if the conversion methods named `_sage_()` shouldn't better be part of Sage? The methods can be added dynamically there. 

Not only is Travis testing a PITA, I have just now several interdependent Sage tickets that add `_sage_` functions, and I need not only patches against SymPy master but also 1.0. And a SymPy patch and a Sage ticket for each.

**Comment by rwst** [comment]
It looks like @certik started the Sage interface in https://github.com/sympy/sympy/commit/98e996ca4b89c6cebcec60f5574621cd8342006e. We certainly would like to hear from him on this.

OTOH the move could be done completely independently in Sage, and I will try to experiment this for principle.

**Comment by certik** [comment]
@rwst I am fine either way. The only disadvantage of adding it in Sage is that if something has to be modified in those `_sage_` methods that is tied to a sympy version, then you have to maintain specific patches to specific SymPy versions. But I guess Sage will use just one version of sympy at a time, so I guess that's fine. How do other Python libraries handle this?

**Comment by rwst** [comment]
Yes, Sage always has one specific SymPy version. I'll now change the issue description.

**Comment by certik** [comment]
I agree that if we don't have to test Sage, it will become simpler. I am just a bit worried that it might happen to be a lot of work to keep things working if things are not tested in SymPy itself. But perhaps it will not be too much work to update the _sage_ methods for each new SymPy release to be put into Sage.

**Comment by rwst** [comment]
I'm afraid the Sage ticket will have to use `_sage_()` instead of `sage()` because all other interfaces in Sage use that too. So it will overwrite what is in SymPy.

**Comment by asmeurer** [comment]
Can you guys test SymPy master in your testing infrastructure so that you aren't surprised by a bunch of accidental breaks every time we do a SymPy release? That's the main disadvantage of moving everything to Sage, is that we can break stuff in SymPy and not really know about it.

**Comment by rwst** [comment]
The ticket is only in part (the last part) about moving the testing. Of relevance to SymPy is testing the interface, and possibly leveraging Sage’s symbolic testsuite. Both are a very limited part of Sage, doctests in a few files that use SymPy. That will increase hopefully. Anyway, to do full Sage testing for this is overkill and not practical in Travis. Maybe it’s possible to have a docker image of a current Sage source installation to run specific tests in Travis. For each PR it would install SymPy and test. The problem is then to provide the images. Also probably maintain a patchset to Sage with changes that are not in current.

As to the Sage testing infrastructure ("patchbot"), I don’t think it easily changeable to use anything other than current Sage plus a specific trac ticket (equivalent to PR), at least not in the distributed official patchbot version because that is about testing trac tickets. If I understand it a Travis testing job is started with every SymPy PR commit. With patchbot it is queued but immedate testing is not guaranteed.

There is interest to include more current SymPy versions in Sage. If the interface is inside Sage it can be tested by Sage testing, Travis did never successfully do that. AFAIK there is no package in Sage where a current master is downloaded. Frequently upgraded packages are Cython, Pynac, and Pari, only the latter had non-release codepoints at some time, but the commit hashes were named specifically in trac tickets and accompanied with patches.

As to breaking stuff, you guys keep SymPy in good shape, the interface can be adapted on upgrade in Sage.

**Comment by ylemkimon** [comment]
@rwst The ticket seems to be closed. Should we remove sage interfaces?

**Comment by rwst** [comment]
The sage interfaces are overwritten in Sage with 8.1.beta10. They were substantially enlarged so the SymPy version can be considered no longer up-to-date. All the tests from `test_sage.py` are now doctests in `sage/interfaces/sympy.py`. I would remove the SymPy version and Travis testing but this is only an opinion, I'm usually quick with such removals.

**Comment by asmeurer** [comment]
I'm still worried that we might break the Sage integration and it won't be caught until after we release (and our release cycle is slow), which will make it hard for Sage to ship the latest SymPy. So I'm not so sure we should remove the testing. In some cases it will be something that we need to fix (or rather, not break) on our part, and in some cases the Sage interfaces in Sage will need to be changed. In the latter case, it's still useful because we can open issues and provide fixes to upstream Sage before SymPy is released (and XFAIL the SymPy test as appropriate). 

If Sage itself is running continuous tests against SymPy master this isn't a big deal, but my understanding is that you don't. I think it's important that SymPy works in Sage and that Sage can use new versions of SymPy as they are released with minimal friction, as Sage is still a fairly important user of SymPy.

However we should definitely remove the `_sage_` functions from SymPy if they are now implemented in Sage.

**Comment by ylemkimon** [comment]
@asmeurer If we remove `_sage_` functions from sympy, we cannot test Sage integration, as SymPy has to be made into a SPKG(Sage package).

**Comment by asmeurer** [comment]
Is it not enough to run the Sage tests with the Sage Python, so that it imports the development SymPy (what I believe it currently does)?

**Comment by ylemkimon** [comment]
@asmeurer I think it's impossible to actually test Sage interfaces on Travis. AFAIK,
(i) some patches are applied to SymPy source codes during installation of SymPy in Sage as SPKG.
(ii) SymPy to Sage conversions(`_sage_ functions) are now dynamically added to SymPy by `sage.interfaces.sympy.sympy_init`
(iii) Sage doctests cannot be tested using Travis(#13395)

**Comment by rwst** [comment]
Yes, to install SymPy master into Sage you need a Sage source installation. The system SymPy is not used by Sage.

**Comment by rwst** [comment]
@asmeurer To install SymPy master into Sage you need a Sage source installation and jump through hoops. The system SymPy is not used by Sage. That's why users rarely try a fresh SymPy with Sage.

**Comment by asmeurer** [comment]
What patches are done to SymPy? Can we upstream them? 

We shouldn't need to install SymPy. It just needs to import the local version instead of the version in Sage (assumedly setting `PYTHONPATH` should do the trick, unless Sage has seriously modified the way Python's import system works).

**Comment by rwst** [comment]
@asmeurer With the upcoming Sage 8.1.beta10 there is only #13476 left as patch. I'll try later if setting `PYTHONPATH` really works.

**Comment by rwst** [comment]
Setting `PYTHONPATH` will not work, because of https://trac.sagemath.org/ticket/22608 --- I tried. What will work is:
```
sage: import sys
sage: sys.path[0]='/usr/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg'
sage: import sympy
sage: sympy.__version__
'1.1.2.dev'
```
Fortunately SymPy is not imported at startup, so unless you use specific calculus functionality of Sage there will be no version clash.

**Comment by rwst** [comment]
About the dynamical addition of `_sage_` methods, this happens starting with 8.1.beta10 in `sage.interfaces.sympy.sympy_init()`. This function is called once when the first Sage object is converted to SymPy---which is suboptimal. This will change with https://trac.sagemath.org/ticket/24067 which patches `sympy.__init__.py` such that the call is made with every import. As you won't have that patch in SymPy proper, just do the call yourself after `import sympy` (see previous comment).

**Comment by asmeurer** [comment]
Why does sage monkeypatch sympy? Is there no way to define a conversion without modifying the objects to be converted?

**Comment by rwst** [comment]
> Why does sage monkeypatch sympy?

 - Ease of maintenance (not having to adapt to another repo/coding standards for a simple two-liner);
 - having it all in one file instead of all over the place (aspect oriented not object oriented);
 - the SymPy/Sage interface is not SymPy, and it is not owned by SymPy; it's also exclusively used by Sage

> Is there no way to define a conversion without modifying the objects to be converted?

Objects are not modified, methods are added to classes. In Aspect-C++ I would have just added an aspect. Without AOP in Python there is only "monkeypatching".

**Comment by ylemkimon** [comment]
So after sympy/sympy#13476 and next Sage release, can we do:
```
sage: import sys
sage: sys.path.insert(0, 'path/to/development_sympy')
sage: from sage.interfaces.sympy import sympy_init
sage: import sympy
sage: sympy_init()
sage: sympy.test('sympy/external/tests/test_sage.py', subprocess=False)
```
?

**Comment by rwst** [comment]
Yes, I just confirmed this with the Sage branch at https://trac.sagemath.org/ticket/24062 which will be part of the next release. Note I didn't succeed when giving the Python path, I had to give the full path to the SymPy egg. To confirm that I really loaded master I checked `sympy.__version__`. You don't need #13476 to pass the test because what is fixed in #13476 was not tested before.

**Comment by ylemkimon** [comment]
Thank you! Then, are #13476 and #13420 no longer needed?

**Comment by asmeurer** [comment]
I mean specifically why not have a method to define conversions that doesn't require monkeypatching (yes, monkeypatching is "modifying")? I would expect the logic to be something like

```
if object defines _sage_:
    call object._sage_
else:
    cal sage_translations[object]
```

where `sage_translations` would be some Sage-level translations (there can also be other layers to this for user defined translations outside of both Sage and SymPy).

How would you handle things for a Python C extension (which generally don't allow monkeypatching)? 

It doesn't really matter from the point of view of this issue. It's just worth noting that monkeypatching generally indicates a bad API design in Python.

**Comment by rwst** [comment]
Agree, a `sage()` function instead of `_sage_()` members would resolve the design problem. One usually choses member functions to get fallbacks in case a subclass doesn't have it implemented but the object tree is not deep here.

**Comment by jdemeyer** [comment]
I'll work on this.

**Comment by jdemeyer** [comment]
I would like to remove the monkey-patching but still implement the conversions in Sage. Essentially, keep the `_sage_()` methods in sympy but make them simple wrappers around functions implemented in Sage. Something like
```python
class SomeSympyClass:
    def _sage_(self):
        from sage.interfaces.sympy import SomeSympyClass_to_sage
        return SomeSympyClass_to_sage(self)
```

**Comment by ylemkimon** [comment]
@rwst AFAIK, sage 8.1 was released few months ago. Could you suggest how we, SymPy, should proceed?

**Comment by rwst** [comment]
> @rwst AFAIK, sage 8.1 was released few months ago. Could you suggest how we, SymPy, should proceed?

Did something change for you? My https://github.com/sympy/sympy/issues/13430#issuecomment-342727945 stands and @jdemeyer made a clear statement in https://github.com/sympy/sympy/issues/13430#issuecomment-349926891.

**Comment by mkoeppe** [comment]
New discussion in https://groups.google.com/g/sympy/c/HwOsl8bFw-c/m/RDresNFfBAAJ

**Comment by oscarbenjamin** [comment]
Having just read this issue I'm unclear what the "issue" is. One interpretation from the above is that SymPy should expunge all `_sage_` methods and then not attempt to test Sage integration...

Currently I don't think there is any Sage integration testing on SymPy's side. When moving from Travis to Actions I tried but it seemed difficult to set up.

**Comment by mkoeppe** [comment]
I'll prepare the PR for Sage integration testing on GH Actions now. When it works, we can test a sympy branch that just expunges all _sage_ methods.

**Comment by certik** [comment]
@mkoeppe thank you for all the work on this. We really appreciate it.

**Comment by oscarbenjamin** [comment]
Yes, thanks indeed for your help with this. It has been something that I wanted to do for a while but hasn't happened because I don't actually know how to do it!

**Comment by mkoeppe** [comment]
This will take about 3.5 hours to run through

**Comment by mkoeppe** [comment]
Fixed in #21958.

**Comment by oscarbenjamin** [comment]
I just merged #21958 so I think this can be closed.
